### ISSUE TITLE
POST Mapping

### ISSUE THREAD

**Issue by flaviocordova** [moleculer-java/moleculer-java-web#5]
Hi, it's me again.. 🙂 

I guess there might be an error processing data when posting to a service through ApiGateway..

I have a service where I have overriden the find action (the action itself is not actually relevant) like this:

```javascript
    find: {
      handler(ctx) {
        console.log('input params:', ctx.params)
        return { status: 'OK' }
      },
    },
```

When I call it from moleculer client, it shows:

```
call connect_skills.find `{"query": { "name": { "$regex": "android", "$options": "i" }}}`
>> Call 'connect_skills.find' with params: { query: { name: { '$regex': 'android', '$options': 'i' } } }
>> Execution time:131ms
>> Response:
{ status: 'OK' }
```

and on service console I see this:

```
input params: { query: { name: { '$regex': 'android', '$options': 'i' } } }
```

So far, so good...

But when I call it using a Post request and passing the same content in the body, it shows a completely different content in ctx.params:

```
input params: Transform {
  _readableState: ReadableState {
    objectMode: false,
    highWaterMark: 16384,
    buffer: BufferList { head: null, tail: null, length: 0 },
    length: 0,
    pipes: [],
    flowing: null,
    ended: false,
    endEmitted: false,
    reading: false,
    sync: false,
    needReadable: false,
    emittedReadable: false,
    readableListening: false,
    resumeScheduled: false,
    errorEmitted: false,
    emitClose: true,
    autoDestroy: true,
    destroyed: false,
    errored: null,
    closed: false,
    closeEmitted: false,
    defaultEncoding: 'utf8',
    awaitDrainWriters: null,
    multiAwaitDrain: false,
    readingMore: false,
    decoder: null,
    encoding: null,
    [Symbol(kPaused)]: null
  },
  _events: [Object: null prototype] { prefinish: [Function: prefinish] },
  _eventsCount: 1,
  _maxListeners: undefined,
  _writableState: WritableState {
    objectMode: false,
    highWaterMark: 16384,
    finalCalled: false,
    needDrain: false,
    ending: false,
    ended: false,
    finished: false,
    destroyed: false,
    decodeStrings: true,
    defaultEncoding: 'utf8',
    length: 0,
    writing: false,
    corked: 0,
    sync: true,
    bufferProcessing: false,
    onwrite: [Function: bound onwrite],
    writecb: null,
    writelen: 0,
    afterWriteTickInfo: null,
    buffered: [],
    bufferedIndex: 0,
    allBuffers: true,
    allNoop: true,
    pendingcb: 0,
    prefinished: false,
    errorEmitted: false,
    emitClose: true,
    autoDestroy: true,
    errored: null,
    closed: false
  },
  allowHalfOpen: true,
  _transform: [Function: transform],
  '$prevSeq': 0,
  '$pool': Map(0) {},
  [Symbol(kCapture)]: false,
  [Symbol(kTransformState)]: {
    afterTransform: [Function: bound afterTransform],
    needTransform: false,
    transforming: false,
    writecb: null,
    writechunk: null,
    writeencoding: null
  }
}
```

**Comment by berkesa** [comment]
Hi! It looks like it sent the data as a Stream. This happens when "Content-Type" is "multipart...".
Can you set the "Content-Type" parameter for the POST request? Like this:
**Content-Type: application/json**
Not 100% that will be so good, but please give it a try. Thanks!

**Comment by flaviocordova** [comment]
Hi @berkesa 

Not, not really.. I'm using postman with application/json.

Remember I'm using a spring filter to redirect the request to apigateway... Could be related or have I missed something?

Just to remember:
```java
@Component
public class ApiGatewayFilter extends GenericFilterBean {
	@Autowired
	private ApiGateway gateway;
	@Autowired
	private ServiceBroker broker;
	
	private AsyncService serviceRunner;
	
	@Override
	protected void initFilterBean() throws ServletException {
		super.initFilterBean();
		this.serviceRunner = new AsyncService(this.broker, this.gateway);
	}
	
	@Override
	public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain) throws IOException, ServletException {
		HttpServletRequest req = (HttpServletRequest) request;
		HttpServletResponse res = (HttpServletResponse) response;
        
		if (!req.getRequestURI().startsWith(CamundaConfiguration.API_PATH + "/mol")) {
    			filterChain.doFilter(request, response);
			return;
		}
    	
		try {
			this.serviceRunner.service(req, res);
		} catch (Exception e) {
			System.out.println("=========== EXCEPTION IN API GATEWAY ==============");
			e.printStackTrace();
			throw new ServletException("API GATEWAY ERROR", e); 
		}
	}
}
```

**Comment by berkesa** [comment]
Hi!
As a test, I called a Node.js-based action via Java-based ApiGateway + REST.
It worked, the data was not transmitted as a Stream (as in the case you described).
I tested it with "pure" ApiGateway + Netty, without Spring MVC.
:( Something else is causing the difference in this operation, maybe the Spring’s FilterBean, or I don’t know.

Suggestions:

1.) Try to add a **RequestLogger** middleware to the route.
2.) Transporters have a **setDebug(true)** function that prints more information that can help to understand the problem.

Please copy the log here with the debug messages, it might make it easier to figure out the difference.
(It's a harder way, but debugging the "AsyncService.java" can also help.)

**Comment by flaviocordova** [comment]
Hi @berkesa , here's the log generated when I post the request.. I deleted some sensititive or noisy data...

```
2020-10-13 09:34:27.959  INFO 57735 --- [nio-8080-exec-1] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring DispatcherServlet 'dispatcherServlet'
2020-10-13 09:34:27.959  INFO 57735 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Initializing Servlet 'dispatcherServlet'
2020-10-13 09:34:28.088  INFO 57735 --- [nio-8080-exec-1] o.s.web.servlet.DispatcherServlet        : Completed initialization in 128 ms
2020-10-13 09:34:28.186  INFO 57735 --- [pool-3-thread-1] s.m.transporter.RedisTransporter         : Submitting message to channel "MOL-development.REQ.accounts_development-Flavios-MacBook-Pro.local_21817":
{
  "ver" : "4",
  "sender" : "commander_development_57735_Flavios-MacBook-Pro.local",
  "id" : "commander_development_57735_Flavios-MacBook-Pro.local:51",
  "action" : "accounts.me",
  "level" : 1,
  "requestID" : "commander_development_57735_Flavios-MacBook-Pro.local:51",
  "params" : {
    "app_session" : "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXNzaW9uX2lkIjoiNWY3ZjZiNDhkYzNiOGE1NTM5YjhhYWRmIiwiZ2VuZXJhdGVkX29uIjoiMjAyMC0xMC0wOFQxOTo0MDo1Ni41NDdaIiwiaWF0IjoxNjAyMTg2MDU2LCJhdWQiOiJodW50ZXJjby5jb20uYnIiLCJpc3MiOiJhY2NvdW50cy5odW50ZXJjby5jb20uYnIifQ.byxziO7ZPlNi7vFWkAtH2YlZOUlo2C0gvH2gmUN1e5A"
  }
}
2020-10-13 09:34:28.656  INFO 57735 --- [ool-2-thread-10] s.m.transporter.RedisTransporter         : Message received from channel "MOL-development.RES.commander_development_57735_Flavios-MacBook-Pro.local":
{
  "id" : "commander_development_57735_Flavios-MacBook-Pro.local:51",
  "meta" : { },
  "success" : true,
  "data" : {
    "_id" : "59edda57b6eb896a9911212b",
    "name": "Flavio Cordova",
    "account_id" : "59edda57b6eb896a9911212b"
  },
  "ver" : "4",
  "sender" : "accounts_development-Flavios-MacBook-Pro.local_21817"
}
2020-10-13 09:34:29.935  INFO 57735 --- [pool-3-thread-1] s.m.transporter.RedisTransporter         : Submitting message to channel "MOL-development.REQ.connect_development-Flavios-MacBook-Pro.local_96299":
{
  "ver" : "4",
  "sender" : "commander_development_57735_Flavios-MacBook-Pro.local",
  "id" : "commander_development_57735_Flavios-MacBook-Pro.local:53",
  "action" : "connect_skills.find",
  "level" : 1,
  "requestID" : "commander_development_57735_Flavios-MacBook-Pro.local:53",
  "params" : {
    "query" : {
      "name" : {
        "$regex" : "android",
        "$options" : "i"
      }
    }
  },
  "meta" : {
    "user" : {
      "account" : {
        "_id" : "59edda57b6eb896a9911212b",
        "name" : "Flavio Cordova",
        "account_id" : "59edda57b6eb896a9911212b",
      }
    }
  },
  "stream" : true,
  "seq" : 0
}
2020-10-13 09:34:31.226  INFO 57735 --- [ool-2-thread-13] s.m.transporter.RedisTransporter         : Message received from channel "MOL-development.RES.commander_development_57735_Flavios-MacBook-Pro.local":
{
  "id" : "commander_development_57735_Flavios-MacBook-Pro.local:53",
  "meta" : {
    "user" : {
      "account" : {
        "_id" : "59edda57b6eb896a9911212b",
        "name" : "Flavio Cordova",
        "account_id" : "59edda57b6eb896a9911212b",
      }
    }
  },
  "success" : true,
  "data" : [ {
    "_id" : "5ee8e0ee8bad4001faca02dc",
    "name" : "Abap"
  } ],
  "ver" : "4",
  "sender" : "connect_development-Flavios-MacBook-Pro.local_96299"
}
Exception in thread "ForkJoinPool.commonPool-worker-3" java.lang.NullPointerException
	at org.apache.tomcat.util.http.NamesEnumerator.findNext(MimeHeaders.java:432)
	at org.apache.tomcat.util.http.NamesEnumerator.nextElement(MimeHeaders.java:458)
	at org.apache.tomcat.util.http.NamesEnumerator.nextElement(MimeHeaders.java:416)
	at services.moleculer.web.servlet.request.AbstractWebRequest$1.next(AbstractWebRequest.java:228)
	at services.moleculer.web.servlet.request.AbstractWebRequest$1.next(AbstractWebRequest.java:1)
	at services.moleculer.web.middleware.RequestLogger$RequestLoggerTask.printRequest(RequestLogger.java:285)
	at services.moleculer.web.middleware.RequestLogger$RequestLoggerTask.run(RequestLogger.java:254)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1429)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)

```

Interestingly, AbstractWebRequest thrown an exception at line 228 and for that to happen, ```req.getHeadersNames()``` should be null, and then I recalled the ThreadSafe thing... Maybe this information is out of reach the current thread, just like the principal.

I was curious though, and I added some messages to my filter to list the header and I got this:

```
---------------------------------
here we go boys...
Processing /cmd/api/mol/cnn/skills/search
Headers we have here:
	connection = close
	content-length = 77
	accept-encoding = gzip, deflate
	cookie = JSESSIONID=94EB629A2A54B8F5BE7CB5628601D958
	host = localhost:8080
	accept = */*
	user-agent = PostmanRuntime/6.4.1
	authorization = Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXNzaW9uX2lkIjoiNWY3ZjZiNDhkYzNiOGE1NTM5YjhhYWRmIiwiZ2VuZXJhdGVkX29uIjoiMjAyMC0xMC0wOFQxOTo0MDo1Ni41NDdaIiwiaWF0IjoxNjAyMTg2MDU2LCJhdWQiOiJodW50ZXJjby5jb20uYnIiLCJpc3MiOiJhY2NvdW50cy5odW50ZXJjby5jb20uYnIifQ.byxziO7ZPlNi7vFWkAtH2YlZOUlo2C0gvH2gmUN1e5A
	content-type = application/json
	postman-token = 8d334e62-58ad-43cb-a1ed-b6ada4a0c021
	cache-control = no-cache
---------------------------------
```

Not only that, after I have iterated the headers myself, RequestLogger stopped throwing the exception and has shown this:
```
2020-10-13 10:16:08.536  INFO 58316 --- [onPool-worker-3] s.m.web.middleware.RequestLogger         : Request from <unknown host> processed within 55 milliseconds (55,720,120 nanoseconds).

  POST /cmd/api/mol/cnn/skills/search HTTP/1.1
  connection: close
  content-length: 77
  accept-encoding: gzip, deflate
  cookie: JSESSIONID=8F5F139449606F61B9FAF8ADB0BE197E
  host: localhost:8080
  accept: */*
  user-agent: PostmanRuntime/6.4.1
  authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzZXNzaW9uX2lkIjoiNWY3ZjZiNDhkYzNiOGE1NTM5YjhhYWRmIiwiZ2VuZXJhdGVkX29uIjoiMjAyMC0xMC0wOFQxOTo0MDo1Ni41NDdaIiwiaWF0IjoxNjAyMTg2MDU2LCJhdWQiOiJodW50ZXJjby5jb20uYnIiLCJpc3MiOiJhY2NvdW50cy5odW50ZXJjby5jb20uYnIifQ.byxziO7ZPlNi7vFWkAtH2YlZOUlo2C0gvH2gmUN1e5A
  content-type: application/json
  postman-token: b24ca657-9f30-4e9a-a4dc-b02e89ef948c
  cache-control: no-cache

  HTTP/1.1 200 OK
  Cache-Control: no-store, no-cache, must-revalidate
  Content-Type: application/json;charset=utf-8
  Content-Length: 15

{"status":"OK"}
```
(I changed the returned data to ```{"status": "OK"}``` in this second test, just to reduce the noise.

However, Moleculer Node is still receiving the stream...

Notice that the transporter logged a ```  "stream" : true,```

**Comment by flaviocordova** [comment]
I found something here at NonBlockingWebRequest

```java
	public NonBlockingWebRequest(ServiceBroker broker, AsyncContext async) throws IOException {
		super((HttpServletRequest) async.getRequest());

		// Create body stream
		if (contentLength != 0) {
			if (multipart) {
				createMultipartStream(broker, async);
			} else {
				createStream(broker, async);
			}
		}
	}
```

Since I have a content-length > 0 and it's NOT multipart, it's considering a stream, isn't it?

How should it be considered when posting an application/json body?

**Comment by flaviocordova** [comment]
I could trace the code to ActionInvoker, line 242.
So far, things look all ok to me... It gets a ```Tree merged = ...``` that is correct and use it to call ```serviceInvoker.call```.

Then I decided to go backwards: The transporter logged stream: true, so I started from it. 
1) Transporter.class adds stream: true when ctx.stream is != null (line 479)
2) Context sets stream if it receives a PacketStream in the constructor (line 99)
3) ActionInvoker will always pass req.getBody() to Context
```java
serviceInvoker.call(new Context(serviceInvoker, eventbus, uidGenerator, uidGenerator.nextUID(),
	actionName, merged, 1, null, null, req.getBody(), opts, nodeID)).then(out -> {
	sendResponse(req, rsp, out);
}).catchError(err -> {
	logger.error("Unable to invoke action!", err);
	sendError(rsp, err);
});
```
4) AbstractWebRequest returns this.stream on getBody(), which is set by ```NonBlockingWebRequest.createStream``` every time it gets a request with a non-multipart request with content-length > 0... So it looks like it will always send any POST as a stream...

**Comment by berkesa** [comment]
Hi @flaviocordova !
Thanks for the feedback and comments, you helped a lot with these. It really was the bug you described.
Fixes for the above are included in version **1.2.11**:
```
group: 'com.github.berkesa', name: 'moleculer-java-web', version: '1.2.11'
```
This version also includes a fix for the RequestLogger's NullPointer bug (and any similar HTTP Header/ThreadLocal errors).
Please write a feedback to let me know if this version works for you.

Some other APIs have also been updated, links need to be modified if you reference them directly in the gradle / maven configuration. The new versions are
moleculer-java: 1.2.15
datatree-core: 1.0.15
datatree-adapters: 1.0.14

**Comment by flaviocordova** [comment]
It did work!!!
Thank you once again!
